extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,617 @@
+import io.druid.indexing.common.task.TaskResource;
+import io.druid.indexing.kafka.KafkaDataSourceMetadata;
+import io.druid.indexing.kafka.KafkaIOConfig;
+import io.druid.indexing.kafka.KafkaIndexTask;
+import io.druid.indexing.kafka.KafkaPartitions;
+import io.druid.indexing.overlord.DataSourceMetadata;
+import io.druid.indexing.overlord.IndexerMetadataStorageCoordinator;
+import io.druid.indexing.overlord.TaskMaster;
+import io.druid.indexing.overlord.TaskStorage;
+import io.druid.indexing.overlord.supervisor.Supervisor;
+import io.druid.indexing.overlord.supervisor.SupervisorSpec;
+import io.druid.metadata.EntryExistsException;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Random;
+import java.util.concurrent.ExecutorService;
+
+/**
+ * Supervisor responsible for managing the KafkaIndexTasks for a single dataSource. At a high level, the class accepts a
+ * {@link KafkaSupervisorSpec} which includes the Kafka topic and configuration as well as an ingestion spec which will
+ * be used to generate the indexing tasks. The run loop periodically refreshes its view of the Kafka topic's partitions
+ * and the list of running indexing tasks and ensures that all partitions are being read from and that there are enough
+ * tasks to satisfy the desired number of replicas. As tasks complete, new tasks are queued to process the next range of
+ * Kafka offsets.
+ */
+public class KafkaSupervisor implements Supervisor
+{
+  /**
+   * Keeps track of the offset range and indexing state of the partitions in our topic. The offset range is only
+   * modified after a task including that partition completes successfully so retries and replicas will always use the
+   * same range. When a new partition is discovered, it will be added with active=false and this partition will only be
+   * included in the next task after a successful run completes. Both of these behaviors are intended to prevent data
+   * duplication since the partition numbers and offset ranges are used in the generation of the sequenceName.
+   */
+  private class PartitionOffsets
+  {
+    long startingOffset;
+    long endingOffset;
+    boolean active;
+
+    public PartitionOffsets(long startingOffset, long endingOffset, boolean active)
+    {
+      this.startingOffset = startingOffset;
+      this.endingOffset = endingOffset;
+      this.active = active;
+    }
+  }
+
+  private class TaskGroup
+  {
+    Map<Integer, PartitionOffsets> partitions = new HashMap<>();
+    Map<String, Optional<TaskStatus>> taskStatuses = new HashMap<>();
+    boolean initialized = false;
+  }
+
+  protected static final String SUPERVISOR_CONTEXT_KEY = "supervisorId";
```

Comment:
It looks like this is just used in the tests, it could be package-private and still work there.